FreeRTOS/FreeRTOS-Plus-TCP
# Before sending a challenge ACK, check the sequence number more properly

Description
-----------
This PR comes in response to issue [#216](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/issues/216): "Possible overflow in xProcessReceivedTCPPacket". @bernd-edlinger : thanks a lot for reporting this.

When a RST packet is received for an existing TCP connection, 3 things can happen:

- The peer knows exactly what seq number the device is expecting: accept the RST and close the socket.
- The peer comes with a seq number that is within the RX window: send a challenge ACK
- The peer doesn't have a clue about the seq number: ignore the RST packet

The check of the sequence number was not totally exact: it would fail if the sequence number is about to pass through zero.
Solution: use the functions `xSequenceGreaterThan()` and `xSequenceLessThan()`. They are defined in FreeRTOS_TCP_WIN.c and now made public:

Test Steps
-----------
It is a bit difficult to do an integration test for this change, it needs a bit of hacking. Start a TCP connection while the sequence number is e.g. `0xFFFFFF00`. Do not send data, but in stead let the peer send a RST packet with a sequence number of e.g. `0x01` or `0x100`.

In stead, I did a simulation:
~~~c
{
    uint32_t ulSequenceNumber, ulCurrentSequenceNumber, ulRxWindowLength;
    bool r1, r2, r3, r4;

    /* A normal situation, no roll-over: */
    ulCurrentSequenceNumber = 0x0000F000U;
    ulSequenceNumber = ulCurrentSequenceNumber + 512;
    ulRxWindowLength = 1460U;
    r1 = xSequenceGreaterThan( ulSequenceNumber, ulCurrentSequenceNumber );
    r2 = xSequenceLessThan( ulSequenceNumber, ( ulCurrentSequenceNumber + ulRxWindowLength ) );
    r3 = ( ulSequenceNumber > ulCurrentSequenceNumber ) ? 1 : 0;
    r4 = ( ulSequenceNumber < ( ulCurrentSequenceNumber + ulRxWindowLength ) ) ? 1 : 0;
    logPrintf("rc = %d %d versus %d %d\n", r1, r2, r3, r4);

    /* Rolling over but within the expected range: */
    ulCurrentSequenceNumber = 0xFFFFFF00U;
    ulSequenceNumber = ulCurrentSequenceNumber + 512;
    ulRxWindowLength = 1460U;
    r1 = xSequenceGreaterThan( ulSequenceNumber, ulCurrentSequenceNumber );
    r2 = xSequenceLessThan( ulSequenceNumber, ( ulCurrentSequenceNumber + ulRxWindowLength ) );
    r3 = ( ulSequenceNumber > ulCurrentSequenceNumber ) ? 1 : 0;
    r4 = ( ulSequenceNumber < ( ulCurrentSequenceNumber + ulRxWindowLength ) ) ? 1 : 0;
    logPrintf("rc = %d %d versus %d %d\n", r1, r2, r3, r4);

    /* Not rolling over, but still within the expected range: */
    ulCurrentSequenceNumber = 0xFFFFFF00U;
    ulSequenceNumber = ulCurrentSequenceNumber + 1;
    ulRxWindowLength = 1460U;
    r1 = xSequenceGreaterThan( ulSequenceNumber, ulCurrentSequenceNumber );
    r2 = xSequenceLessThan( ulSequenceNumber, ( ulCurrentSequenceNumber + ulRxWindowLength ) );
    r3 = ( ulSequenceNumber > ulCurrentSequenceNumber ) ? 1 : 0;
    r4 = ( ulSequenceNumber < ( ulCurrentSequenceNumber + ulRxWindowLength ) ) ? 1 : 0;
    logPrintf("rc = %d %d versus %d %d\n", r1, r2, r3, r4);
}

/*
    The output is:
    rc = 1 1 versus 1 1 // good
    rc = 1 1 versus 0 1 // wrong
    rc = 1 1 versus 1 0 // wrong
    The results on the left hand are all correct.
*/
~~~


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
